src/interface/sparse_pack_base.hpp

```diff
@@ -0,0 +1,199 @@
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_SPARSE_PACK_BASE_HPP_
+#define INTERFACE_SPARSE_PACK_BASE_HPP_
+
+#include <algorithm>
+#include <functional>
+#include <limits>
+#include <map>
+#include <memory>
+#include <regex>
+#include <string>
+#include <tuple>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "coordinates/coordinates.hpp"
+#include "interface/variable.hpp"
+#include "utils/utils.hpp"
+
+namespace parthenon {
+namespace impl {
+struct PackDescriptor {
+  std::vector<std::string> vars;
+  std::vector<bool> use_regex;
+  std::vector<MetadataFlag> flags;
+  bool with_fluxes;
+  bool coarse;
+
+  PackDescriptor(const std::vector<std::string> &vars, const std::vector<bool> &use_regex,
+                 const std::vector<MetadataFlag> &flags, bool with_fluxes, bool coarse)
+      : vars(vars), use_regex(use_regex), flags(flags), with_fluxes(with_fluxes),
+        coarse(coarse) {
+    PARTHENON_REQUIRE(use_regex.size() == vars.size(),
+                      "Must have a regex flag for each variable.");
+    PARTHENON_REQUIRE(!(with_fluxes && coarse),
+                      "Probably shouldn't be making a coarse pack with fine fluxes.");
+  }
+
+  PackDescriptor(const std::vector<std::pair<std::string, bool>> &vars_in,
+                 const std::vector<MetadataFlag> &flags, bool with_fluxes, bool coarse)
+      : flags(flags), with_fluxes(with_fluxes), coarse(coarse) {
+    for (auto var : vars_in) {
+      vars.push_back(var.first);
+      use_regex.push_back(var.second);
+    }
+    PARTHENON_REQUIRE(!(with_fluxes && coarse),
+                      "Probably shouldn't be making a coarse pack with fine fluxes.");
+  }
+
+  PackDescriptor(const std::vector<std::string> &vars_in,
+                 const std::vector<MetadataFlag> &flags, bool with_fluxes, bool coarse)
+      : vars(vars_in), use_regex(vars_in.size(), false), flags(flags),
+        with_fluxes(with_fluxes), coarse(coarse) {
+    PARTHENON_REQUIRE(!(with_fluxes && coarse),
+                      "Probably shouldn't be making a coarse pack with fine fluxes.");
+  }
+};
+} // namespace impl
+
+using namespace impl;
+
+class PackIdx {
+  std::size_t vidx;
+  int off;
+
+ public:
+  KOKKOS_INLINE_FUNCTION
+  explicit PackIdx(std::size_t var_idx) : vidx(var_idx), off(0) {}
+  KOKKOS_INLINE_FUNCTION
+  PackIdx &operator=(std::size_t var_idx) {
+    vidx = var_idx;
+    off = 0;
+    return *this;
+  }
+  KOKKOS_INLINE_FUNCTION
+  PackIdx(std::size_t var_idx, int off) : vidx(var_idx), off(off) {}
+  KOKKOS_INLINE_FUNCTION
+  std::size_t Vidx() { return vidx; }
```

Comment:
Why not use `operator()` here instead of `Vidx()`?

Reply:
To me, it is not obvious exactly what `operator()` should return for a `PackIdx`. It could be interpreted to be the variable index as you suggest, but I think that opens up the possibility of confusion. Since it should not be called anywhere in downstream code (so a longer name won't make things much more verbose), I think it is better to have a descriptive name. I have changed it to `VariableIdx()` for better clarity.

Reply:
:+1: